### DISCUSSION TITLE
Support for org.apache.cxf.jaxrs.ext.multipart

### DISCUSSION THREAD

**Discussion by languitar** [quarkusio/quarkus#22002] [Q&A]
Is there any reason why the resteasy REST client uses its own data types for multipart handling and not the types and annotations provided in `org.apache.cxf.jaxrs.ext.multipart`? Or is there any chance to get support for these types? By default, they are generated by the OpenAPI generator target jax-rs and changing the templates to generate code compatible with the Quarkus-specific implementation is not the easiest task there.

**Comment by michalszynkiewicz** [comment]
@languitar Quarkus REST Clients are not based on CXF, why would they use CXF classes?
Aoache CXF is an alternative library for JAX-RS (and much more)

**Reply by maxandersen** [reply]
I think the issue is that op are trying to use openapi generation and haven't found an implementation that works with resteasy and thus when going outside spec apis which a client implementation might to the mismatch Happens. 

Do I have that right op?

**Reply by languitar** [reply]
Yes, I am trying to use the OpenAPI generator and the `microprofile` target there somehow uses CXF as the target for multipart handling.

**Reply by Postremus** [reply]
Just in case it helps, this is my configuration for the OpenAPI generator.
This generates all necessary model classes.
The generated client interface contains only jaxrs annotations, and can be used in combination with the RestClientBuilder (from quarkus-rest-client).


```
<plugin>
				<groupId>org.openapitools</groupId>
				<artifactId>openapi-generator-maven-plugin</artifactId>
				<version>4.3.1</version>
				<executions>
					<execution>
						<id>generate-client</id>
						<phase>generate-sources</phase>
						<goals>
							<goal>generate</goal>
						</goals>
						<configuration>
							<inputSpec>path-to-openapi.yaml</inputSpec>
							<apiPackage>packagename.api</apiPackage>
							<modelPackage>packagename.model</modelPackage>
							<generatorName>jaxrs-spec</generatorName>
							<configHelp>false</configHelp>
							<skipValidateSpec>true</skipValidateSpec>
							<configOptions>
								<java8>true</java8>
								<dateLibrary>java8-localdatetime</dateLibrary>
								<generatePom>false</generatePom>
								<interfaceOnly>true</interfaceOnly>
								<useRuntimeException>true</useRuntimeException>
								<useSwaggerAnnotations>false</useSwaggerAnnotations>
								<useBeanValidation>false</useBeanValidation>
								<hideGenerationTimestamp>true</hideGenerationTimestamp>
							</configOptions>
						</configuration>
					</execution>
				</executions>
			</plugin>
```

Interface looks like this:
```
package packagename.api;

import org.jboss.resteasy.annotations.jaxrs.PathParam;
import org.jboss.resteasy.annotations.jaxrs.QueryParam;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import java.util.Set;

@Path("/extensions")
public interface ExtensionsService {

    @GET
    @Path("/stream/{stream}")
    Set<Extension> getByStream(@PathParam String stream, @QueryParam("id") String id);
}
```

Usage:
```
public  Set<Extension> getExtensions() {
  RestClientBuilder builder = RestClientBuilder.newBuilder().baseUri(<base-uri>);
  ExtensionsService extAPi = builder.build(ExtensionsService .class);
  try {
    return extAPi.getByStream(null, null);
  } finally {
    ((Closeable)extAPi).close();
  }
}
```

**Reply by languitar** [reply]
Isn't jaxrs-spec a server stub generator?

**Reply by Postremus** [reply]
Yes, it is a server-stub generator.
With `<interfaceOnly>true</interfaceOnly>` however, no server specific parts (like resources) are generated. All annotations are from the jaxrs api.
The generated client interface can simply be used by the RestClientBuilder, as shown above.

**Reply by languitar** [reply]
Ok, interesting idea. What does this generate for POST routes with multipart uploads?

In our case, we generate Maven packages from the OpenAPI specification for reuse in different microservices. Therefore, the generator needs to be able to create a working project that can be packaged and pushed into a registry. I suspect that doesn't work with the jaxrs-spec generator in a suitable way?

**Reply by Postremus** [reply]
For multipart uploads, this won't work. Multipart is not part of the jaxrs spec (afaik).
Yes, this won't work for your case, since the generated "client" is not usable without requiring extra code.

The micrprofile templates also only include cxf because of multipart support.
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Java/libraries/microprofile/api.mustache#L15

Another idea, you could write your own openapi-generator templates, based on the micrprofile ones.
These templates could then use e.g. the multipart support from resteasy. https://quarkus.io/guides/rest-client-multipart

**Reply by languitar** [reply]
> Another idea, you could write your own openapi-generator templates, based on the micrprofile ones.
These templates could then use e.g. the multipart support from resteasy. https://quarkus.io/guides/rest-client-multipart

I've tried that. But that's pretty hard to achieve with only template modifications. By default, form models are not generated and the limited mustache templating makes it close to impossible to generate something working without also implementing a custom code generator java class. So not exactly an easy route to go.

**Reply by michalszynkiewicz** [reply]
There's some work on a built in code generator for openapi linked to this issue https://github.com/quarkusio/quarkus/issues/16401
This extension is where we should make it work easily.

If CXF can put their own custom stuff into the generator maybe we could too, hidden by some switch? I'm not very familiar with the project, unfortunately.

**Reply by languitar** [reply]
I've tried that and I've also seen that issue. It will probably be possible by providing a custom generator class. Only templating changes are probably not sufficient or at least hard to maintain due to the fact that resteasy multipart requires form data to be a model class. Typical java generators treat form parameter similar to query or path parameters with individual arguments per endpoint method. Getting rid of this behavior is a bit cumbersome.
